File: code/src/test/java/chapter1/java2/SessionsTests.java
package chapter1.java2;

import chapter1.kotlin13.Sessions;
import chapter1.kotlin8.Session;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Optional;

import static java.util.Collections.emptyList;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class SessionsTests {

    private List<Session> sessions = List.of(
        new Session("Title", emptyList()),
        new Session("Title Too", emptyList())
    );

    @Test
    public void find_session_with_title() {
        assertEquals(
            Optional.of(sessions.get(1)),
            Sessions.findWithTitle(sessions, "Title Too")
        );
    }

    @Test
    public void find_no_session_with_title() {
        assertEquals(
            Optional.empty(),
            Sessions.findWithTitle(sessions, "No such")
        );
    }

    /// begin: baz
    @Test
    public void find_in_no_sessions() {
        assertEquals(
            Optional.empty(),
            Sessions.findWithTitle(emptyList(), "No such")
        );
    }
    /// end: baz
}

File: code/src/test/java/chapter1/kotlin2/SessionsTests.kt
package chapter1.kotlin2

import chapter1.kotlin14.Sessions.findWithTitle
import chapter1.kotlin14.Sessions.firstWithTitle
import chapter1.kotlin14.Sessions.toOptional
import chapter1.kotlin8.Session
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.util.*
import java.util.List

class SessionsTests {
    private val sessions = List.of(
        Session("Title", emptyList()),
        Session("Title Too", emptyList())
    )

    /// begin: foo
    @Test
    fun find_session_with_title() {
        Assertions.assertEquals(
            Optional.of(sessions[1]),
            sessions.firstWithTitle("Title Too").toOptional() /// change: foo
        )
    }
    /// end: foo
    
    /// begin: bar
    @Test
    fun `find session with title`() {
        assertEquals(
            sessions[1],                         ///  change: bar
            sessions.firstWithTitle("Title Too") ///  change: bar
        )
    }
    /// end: bar

    /// begin: baz
    @Test
    fun find_no_session_with_title() {
        Assertions.assertEquals(
            Optional.empty<Any>(),
            sessions.findWithTitle("No such")
        )
    }
    /// end: baz

    @Test
    fun find_in_no_sessions() {
        Assertions.assertEquals(
            Optional.empty<Any>(),
            emptyList<Session>().findWithTitle("No such")
        )
    }
}